internal/impl/elasticsearch/output_v2.go

```diff
@@ -0,0 +1,436 @@
+		}
+		batchErr.Failed(i, err)
+	}
+
+	batchInterpolator := eso.newInterpolationExecutor(&batch)
+
+	indexer, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+		Client: eso.client,
+		OnError: func(ctx context.Context, err error) {
+			eso.log.Errorf("elasticsearch_v2: Bulk Indexer Error: %s\n", err)
+		},
+		Timeout: eso.conf.timeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	for i, msg := range batch {
+		index, action, id, routing, err := batchInterpolator.exec(i)
+		if err != nil {
+			eso.log.Errorf("Failed to execute bloblang interpolation: %v\n", err)
+			batchErrFailed(i, err)
+			continue
+		}
+
+		msgBytes, err := msg.AsBytes()
+		if err != nil {
+			eso.log.Errorf("Failed to get message bytes: %v\n", err)
+			batchErrFailed(i, err)
+			continue
+		}
+
+		switch action {
+		case "update", "upsert":
+			updateBody := map[string]any{
+				"doc":           json.RawMessage(msgBytes),
+				"doc_as_upsert": action == "upsert",
+			}
+			updateBodyBytes, err := json.Marshal(updateBody)
+			if err != nil {
+				eso.log.Errorf("Failed to marshal update/upsert doc: %v\n", err)
+				batchErrFailed(i, err)
+				continue
+			}
+			err = indexer.Add(ctx, esutil.BulkIndexerItem{
+				Index:      index,
+				Action:     "update",
+				DocumentID: id,
+				Routing:    routing,
+				Body:       bytes.NewReader(updateBodyBytes),
+				OnSuccess:  onSuccessHandler(eso),
+				OnFailure:  onFailureHandler(eso, i, batchErrFailed),
+			})
+			if err != nil {
+				eso.log.Errorf("Failed to add message to elasticsearch indexer: %v\n", err)
+				batchErrFailed(i, err)
+			}
+		case "delete":
+			err = indexer.Add(ctx, esutil.BulkIndexerItem{
+				Index:      index,
+				Action:     action,
+				DocumentID: id,
+				Routing:    routing,
+				OnSuccess:  onSuccessHandler(eso),
+				OnFailure:  onFailureHandler(eso, i, batchErrFailed),
+			})
+			if err != nil {
+				eso.log.Errorf("Failed to add message to elasticsearch indexer: %v\n", err)
+				batchErrFailed(i, err)
+			}
+		case "index", "create":
+			err = indexer.Add(ctx, esutil.BulkIndexerItem{
+				Index:      index,
+				Action:     action,
+				DocumentID: id,
+				Routing:    routing,
+				Body:       bytes.NewReader(msgBytes),
+				OnSuccess:  onSuccessHandler(eso),
+				OnFailure:  onFailureHandler(eso, i, batchErrFailed),
+			})
+			if err != nil {
+				eso.log.Errorf("Failed to add message to elasticsearch indexer: %v\n", err)
+				batchErrFailed(i, err)
+			}
+		default:
+			batchErrFailed(i, fmt.Errorf("elasticsearch action '%s' is not allowed", action))
+		}
+	}
+
+	err = indexer.Close(ctx)
+	if err != nil {
+		return err
+	}
+
+	if batchErr != nil && batchErr.IndexedErrors() > 0 {
+		return batchErr
+	}
+
+	return nil
+}
+
+func (eso *EsOutput) Close(context.Context) error {
+	return nil
+}
+
+//------------------------------------------------------------------------------
+
+func (eso *EsOutput) newInterpolationExecutor(batch *service.MessageBatch) (ie interpolationExecutor) {
```

Comment:
Does this need to be a pointer?
```suggestion
func (eso *EsOutput) newInterpolationExecutor(batch service.MessageBatch) (ie interpolationExecutor) {
```

Reply:
@jem-davies A slice already contains a pointer to the underlying array -- so am unsure what the purpose of using a reference to a struct that already contains a reference is

Reply:
Basically a slice has 3 fields: a pointer to the underlying array, a length field, and a capacity field. We wont be passing the entire array each time, just the 3 values mentioned before. 

I think that any performance benefit we get from passing a pointer to a slice is offset by needing to extract the pointer in the first place.

In anycase, it's not _really_ idiomatic and there are so few fields I don't think performance needs to be a consideration here 🤷